repair/repair.cc

```diff
@@ -1658,9 +1684,25 @@ future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_me
     using inet_address = gms::inet_address;
     return seastar::async([&db, tm = std::move(tm), tokens = std::move(bootstrap_tokens)] () mutable {
         auto keyspaces = db.local().get_non_system_keyspaces();
-        rlogger.info("bootstrap_with_repair: started with keyspaces={}", keyspaces);
         auto myip = utils::fb_utilities::get_broadcast_address();
         auto reason = streaming::stream_reason::bootstrap;
+        // Calculate number of ranges to sync data
+        size_t nr_ranges_total = 0;
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            dht::token_range_vector desired_ranges = strat.get_pending_address_ranges(tm, tokens, myip);
+            seastar::thread::maybe_yield();
+            nr_ranges_total += desired_ranges.size();
+        }
+        db.invoke_on_all([nr_ranges_total] (database&) {
+            _node_ops_metrics.bootstrap_finished_ranges = 0;
+            _node_ops_metrics.bootstrap_total_ranges = nr_ranges_total;
+        }).get();
```

Comment:
I think we should let those keep incrementing. So total_ranges += nr_ranges_total. This way you can look at the history of the metric and see where repairs are initiated and how they complete.

Another thing to consider is the unit. Do we want it to be ranges? Perhaps we should define "1" as the entire ring, so total is incremented by (sum of vnodes / total ring size) and finished is incremented by (range size in tokens / total ring size). (total ring size = 2^64). This normalizes the work to units that are independent of implementation details like ranges.

Reply:
> I think we should let those keep incrementing. So total_ranges += nr_ranges_total. This way you can look at the history of the metric and see where repairs are initiated and how they complete.

Keep increasing the metrics has the problem of one operation mess up with previous ones. For example, first rebuild needed to finish 100 ranges, but it finished only 50 due to errors. The second rebuild needed to finish 200 and finished 200. The always incrementing finished and total number will be wrong.  Well we can increase finished number for the first rebuild when it fails, but we can not tell if it is really finished successfully or not, plus we have to maintain very carefully otherwise the number will be wrong for all following operations. 

To track the history, we can add another metrics. To track progress, I think it is better to "gauge" per operation.

Reply:
> Another thing to consider is the unit. Do we want it to be ranges? Perhaps we should define "1" as the entire ring, so total is incremented by (sum of vnodes / total ring size) and finished is incremented by (range size in tokens / total ring size). (total ring size = 2^64). This normalizes the work to units that are independent of implementation details like ranges.

The progress = finished / total. The total ring size cancels with each other in the equation. I think number of "ranges" is much easier for user to reason about than the sum of length of of ranges divided by the total ring length:  (len(r1) + len (r2) + ...  len(rn))  / 2^64. 

We have plenty of ranges information in the log too, e.g., in streaming and repair "x out of y  ranges".

At some point, if we ever want to drop the concept of ranges, we can revise those metrics.

Reply:
If we make it a gauge then that's fine - we don't need finished/total, just progress.

But let's count tokens, not ranges. So it's tokens_finished/tokens_total. If we have multiple repairs in parallel, it's sum(tokens_finished in each active repair) / sum(tokens_total in each active repair).

Reply:
Counting tokens do not give any more information or make the result more accurate. For example, we have 3 ranges to work on.  [10,20], [30,40], [50,70].  At he the point the first two ranges are finished:

Th progress counting tokens = 20 - 10 + 40 - 30 /   20 - 10 + 40 - 30 + 70 - 50  =  20 / 40 = 50%.
The progress counting ranges = 2 / 4 = 50%.

Why do we even bother to calculate the length of the ranges (counting by tokens)?


With multiple repair jobs, we can still use sum(ranges finished in each active repair) / sum (ranges total in each active repair).

Reply:
Ranges are meaningless for the user, because their size is unclear.

If we divide ranges, they cancel out, but it only works if the ranges are of the same size. In that case it's the same as dividing tokens: completed_ranges/total_tanges = (completed_ranges*tokens_per_range)/(total_ranges*tokens_per_range) = completed_tokens/total_tokens.

Reply:
I think ranges do have meaning. It presents how many fragments of the token ranges the operations have to work on. The progress is a rough estimation anyway. Even if we use count len(range.end() - range.start()), it is still a rough estimation because of the actual data distribution. 

If we agree we only provide the progress metric (as percentage), it makes the argument to count the len(range.end() - range.start()) or number of ranges even less important, because only the percentage is exposed to user. 

Btw, we do not have the "length"  in range class. 

```
510     const optional<bound>& start() const {
511         return _interval.start();
512     }
513     const optional<bound>& end() const {
514         return _interval.end();
515     }
```

Reply:
You are right about the same size thing. Another example [10,20], [30,40], [50,50000] shows it clearly.

Reply:
We don't have length in intervals because intervals might not be numberic (and might be open). You have to extract the bounds and compute them yourself.

Alternatively we can add an optional member function with constraints, see https://godbolt.org/z/1xKc89.

Reply:
Considering the little gain of computing "length" v.s. number of ranges. I still think it is better and good enough to use ranges_finished.size() / ranges_total.size(). It is simpler: it is O(1), no need to calculate "length" of all ranges. No dependence on the interval computing. 

What do you think?

Reply:
Depends on the variation between range sizes. If there is high variation, we will get poor progress tracking.

If there is low variation, it is good enough.

Reply:
The real progress depends on the data distribution over those ranges as well. This data distribution can have very big variation too. Considering this factor, why not keep it simple. 


Btw, Avi, do you insist we should only expose the progress percentage through the metrics or it is fine to expose both finished and total?

Reply:
> The real progress depends on the data distribution over those ranges as well. This data distribution can have very big variation too. Considering this factor, why not keep it simple.

Usually a non-tiny token range has good distibution. If it doesn't, we can't do anything about it. If it does, let's be as accurate as we can be.

> 
> Btw, Avi, do you insist we should only expose the progress percentage through the metrics or it is fine to expose both finished and total?

I dislike percentages and prefer exposing more direct metrics, and to do the percentage computation on the client side. So I prefer finished/total, provided we can get reasonable values for them when repair is not running.

Reply:
Do you insist on we must compute "length" of a range than size of the range (ranges.size())? I can give you another example. Say we have 2 keyspaces, one has finished 9 out of 10 ranges, another also has finished 1 out of 10 ranges. But one has 1GB of data while the other has 1TB of data. We will get progress =  9 + 1 / 10 + 10 = 50%. Apparently, it hasn't finished 50% of data yet. We will get similar 50% progress even if we compute "lengh" of a range.

Reply:
> I dislike percentages and prefer exposing more direct metrics, and to do the percentage computation on the client side. So I prefer finished/total, provided we can get reasonable values for them when repair is not running.

For regular repair, in the latest version, I am combining all ongoing repair jobs. When a repair job is finished it will be removed from the calculation. With percentage, it  gives "1" when there is no repair running. With finish/total, we will return 0 and 0.

Reply:
For node operations, when it is finished, we keep the finish/total value, so the client can sample it correctly. 

Exposing percentage has the advantage of hiding the slight differences between regular and node operations and unit of ranges.

Reply:
But the graph looks bad. Bad looking graphs are not the most important motivation, but if we can find a good way to expose the information and have a nice graph, we should do it.